New cli

hildjj opened the pull request:
Split into several chunks:

- Compile the existing CLI with rollup
- Add tests to the exisitng CLI
- Move the CLI to[ commander](https://github.com/tj/commander.js#readme), updating tests
- Add -t/-T
- Fix #161 by adding support for .js/.cjs config files and being careful about the command line overriding the config file, but the config file overriding defaults.

hildjj commented:
This needs some testing on windows before it's ready.

Mingun commented on bin/peggy.mjs:
```diff
@@ -0,0 +1,331 @@
+  } else {
+    outputStream = fs.createWriteStream(outputFile);
+    outputStream.on("error", () => {
+      abort("Can't write to file \"" + outputFile + "\".");
```
```suggestion
      abort(`Can't write to file "${outputFile}".`);
```

hildjj replied:
Willfix, along with other comments I've marked as "resolved" without commenting on them.  I wonder if we should add `prefer-template`to the lint file.

Mingun commented on bin/peggy.mjs:
```diff
@@ -0,0 +1,331 @@
+    });
+  }
+
+  outputStream.write(source);
```
Currently, that will fail/write the wrong results if `--test/--test-file` option is supplied

hildjj replied:
I think it works as expected.  If you specify a test using either mechanism, and you don't specify an output file, you get the output of running the parser on stdout, colorized if stdout is a tty.  If you specify an output file, that's where the output goes.

Mingun replied:
Yes, now I see. I was think that `source` is the parser object at this moment. Would be not what you expect, however. I would expect that `--test` is fully additive, so parser is still generated and writted to the `--output`, and additionally tested against provided input. Instead when the test flag is active, the parser generated only in-memory and test results is written to the `--output`. Not an obvious behaviour if you ask me.

I would suggest the following behaviour:
- if `--output` is explicitly defined, dump parser code to it, even if `--output` is stdout
- when `--test` is defined, the generated parser `eval`uated and tested against input
- test results always written to the stdout
- if `--output` is not explicitly defined and `--test` is defined, do not output the parser source to stdout, but only test results

So, if you just want to test your just-now-written parser, you use only the `--test` option:
```console
$ peggy --test "1" "foo = '1' { return 'success'; }"
success
```
If you want to store parser into file and test in immediately, you specify  the `--output` and `--test` options:
```console
$ peggy --test "1" --output parser.js "foo = '1' { return 'success'; }"
success
...parser is saved to the `parser.js`...
```
if you want to generate parser and do not want to test it and also do not want to write it to the file, you does not specify neither `--output` nor `--test` options:
```console
$ peggy "foo = '1' { return 'success'; }"
...parser code...
```

Mingun commented on rollup.config.js:
```diff
@@ -29,6 +29,37 @@ const umd_config = {
+    format : "cjs",
+    preferConst: true,
+    exports: "auto",
+    banner: "#!/usr/bin/env node",
```
I think that adding a note about that this file is auto-generated also will be useful

Mingun commented on tsconfig.json:
```diff
@@ -34,5 +34,5 @@
     // "allowSyntheticDefaultImports": true,        /* Allow default imports from modules with no default export. This does not affect code emit, just typechecking. */
   },
-  "include": ["./lib/**/*.js", "./test/**/*.spec.js", "./test/**/*.ts", "./benchmark/*.js"]
+  "include": ["./lib/**/*.js", "./test/**/*.spec.js", "./benchmark/*.js"]
```
This is because we does not need to convert test files to JS to run them?

hildjj replied:
Correct.  Jest handles this.  Keeping the CLI tests as .ts means they won't be rolled up into the web tests.

Mingun replied:
Understood

Mingun commented on bin/peggy.mjs:
```diff
@@ -0,0 +1,331 @@
+
+    abortError("Error parsing JSON:", e);
+  }
+  if ((extraOptions === null)
+      || (typeof extraOptions !== "object")
```
Aren't extra brackets prohibited by the linting rules?
```suggestion
  if (extraOptions === null
      || typeof extraOptions !== "object"
```

hildjj replied:
No.

https://github.com/peggyjs/peggyjs-eslint-config/blob/46d7fcef3477325971c9c79a5c49717d960bc1cd/index.js#L59

I'm open to changing it, but I often use lots of extra parens so that the reader doesn't have to know as much about operator precedence.

Mingun replied:
Ok. I'm personally tend to not put extra braces around comparison operators, but that not a blocker

Mingun reviewed (changes requested):
In general very good, but test file should be read after the parser generation and changing output type from `source` to `parser` can lead to some unintended results. I think, that this should be fixed.

The other stuff are minor but it will be great if you look at them.

Mingun commented on test/cli/run.spec.ts:
```diff
@@ -0,0 +1,388 @@
+      ...process.env,
+      ...opts.env,
+    };
+    const c = spawn(bin, opts.args, {
```
Tests on windows failed because this line executes script using shebang magic, which not work under Windows. This should work:
```suggestion
    if (opts.args !== undefined) {// strange, just check if (opts.args) is not enough
      opts.args.unshift(bin);
    } else {
      opts.args = [bin];
    }
    const c = spawn("node", opts.args, {
```

See https://stackoverflow.com/questions/43419893/how-do-i-fix-error-spawn-unknown-with-node-js-v7-8-0-on-windows-10

hildjj replied:
I did this if `process.platform === "win32"`, which lets me test both ways.

Mingun replied:
Ok

Mingun commented on test/cli/run.spec.ts:
```diff
@@ -0,0 +1,388 @@
+  it("uses dash-dash", async() => {
+    await expect(exec({
+      args: ["--", "--trace"],
+    })).rejects.toThrow("no such file or directory, open '--trace'");
```
Strange, but on Windows the error contains full path to the file, so changed to regexp
```suggestion
    // On Windows error contains the full path to the file: <work_dir>\--trace
    })).rejects.toThrow(/no such file or directory, open '[^']*--trace'/);
```

Mingun reviewed (changes requested):
This two changes fixes the Windows build

o-alexandre-felipe commented on bin/peggy.mjs:
```diff
@@ -0,0 +1,331 @@
+  delete options.format;
+
+  // Command line overwrites config file
+  options = Object.assign({}, options.extraOptionsFile, options);
```
Yow would you expect it to behave with
``` js
options.extraOptions = {a:{b:true, c: true}};
options = {a: {b:false}}
```
?

Mingun commented on rollup.config.js:
```diff
@@ -29,6 +29,37 @@ const umd_config = {
   ],
 };
 
+const cli_config = {
```
[The rollup documentation](https://rollupjs.org/guide/en/#config-intellisense) suggests using or JSDoc comment, or `defineConfig` helper to leverage IDE's Intellisense

Mingun commented on rollup.config.js:
```diff
@@ -29,6 +29,37 @@ const umd_config = {
 
+const cli_config = {
+  onwarn(message) {
+    if (message.code === "EVAL") { return; }
```
Why this needed? If I understand correctly, this code suppresses warnings about `eval(...)` usage, but `eval` not used in the CLI code. I tried to build with this line commented and did not notice any warnings.

It is better to add a comment describing what exactly happens here

hildjj replied:
Removed for the CLI.

hildjj commented:
Sorry it's taken so long, but I'm finally addressing all of these comments.

@mingun, you said "changing output type from source to parser can lead to some unintended results".  Can you say more about that?  My thought was that if you're supplying a test, we should get a parser back instead of source, so that we can call that generated code.

hildjj commented on bin/peggy.mjs:
```diff
@@ -0,0 +1,331 @@
+  )
+  .option("-o, --output <file>", "output file")
+  .option(
+    "--plugin <module>",
```
I don't think this is as clean as just doing `--plugin foo --plugin bar`, because otherwise you'll often need to use `--` to stop the list of plugins.

Mingun replied:
According to examples you will need the stop marker only if plugins were the last option in the arguments list. If you have additional flags after it the stopper will no required, i.e. you can write:
```sh
peggy --plugin a b c -с options.json grammar.peggy
```
and that will work. However this is still a matter of personal preferences, so you can leave it as is

hildjj commented on bin/peggy.mjs:
```diff
@@ -0,0 +1,331 @@
+let options = program
+  .version(peg.VERSION, "-v, --version")
+  .arguments("[input_file]")
+  .addOption(
+    new Option(
+      "--allowed-start-rules <rules>",
+      "comma-separated list of rules the generated parser will be allowed to start parsing from"
+    )
+      .default([], "the first rule in the grammar")
+      .argParser((val, prev) => prev.concat(val.split(",").map(x => x.trim())))
+  )
```
Agree that this should match the others.  If we leave the others as they currently are, we can just remove "comma-separated" from the docs, and the implementation should work ok.

Mingun commented on bin/peggy.mjs:
```diff
@@ -0,0 +1,331 @@
+    "make generated parser cache results"
+  )
+  .option(
+    "-d, --dependency <dependency>",
```
Ok

Mingun commented on bin/peggy.mjs:
```diff
@@ -0,0 +1,341 @@
+
+if (options.extraOptions) {
+  Object.assign(options, options.extraOptions);
+  delete options.extraOptions;
```
I see potential problem with that deletion. If custom options define the `extraOptions` key, it will be deleted here. Need to delete it before doing
```js
Object.assign(options, options.extraOptions);
```

hildjj commented:
>  I would expect that --test is fully additive, so parser is still generated and written to the --output

I can see this point of view.  There have been a few times where I've wanted to save the output of the test, but I suppose I can always just capture stdout or add another option later if it becomes important.

hildjj commented:
After thinking about it some more, I don't want to add variadic options because it would break backward-compatibility for some uses of the command line.

Mingun commented:
> After thinking about it some more, I don't want to add variadic options because it would break backward-compatibility for some uses of the command line.

Yes, I also thinking about this after you write this . So may be later.

I think it is almost done, but I don't like the way how the options combined from the config file, command line JSON and command line parameters -- it is too sophisticated and non-obvious. I suggest to change handling, as I did in https://github.com/Mingun/peggy/pull/new/new-cli (and also fix couple of other issues).

Also, If I understood correctly, do you agree with me that it is worth reworking the way the `--test` is handled?

hildjj commented:
I have reworked how `--test` behaves, as discussed above.  I'm just fixing up the tests now before pushing the update. 

I agree that the config file stuff is a little subtle, but I think it was always that way.  Can you explain a little more about how you changed it?  I'm not seeing how yours is different from mine in the link above.

Mingun commented:
Conceptually I did the following:
1. I introduce `DEFAULTS` const with all options that have a default value
2. Then I save parsed config from the commander to the `cliOptions`.
2. Then remove from `cliOptions` all keys that have default values (as defined in the `DEFAULTS`)
3. Then build new `options` object as
```js
Object.assign({}, DEFAULTS, extraOptionsFile, extraOptions, cliOptions)
```

hildjj commented:
That makes sense.  I'll work in something like that.

hildjj commented:
As a compromise on variadic options, and to make everything consistent, you can now specify a comma-separated list for `--dependency` and `--plugin`

Mingun commented:
As for dancing around config, I've just create an issue in the commander, so, if it will be implemented, we could switch to the built-in solution: https://github.com/tj/commander.js/issues/1584

hildjj commented:
I think I have all of the requested changes in now.  The defaults handling is better than what I had before, but wasn't able to be as clean as you had, because I wanted arrays and objects in the options to be additive instead of replacing what the earlier option was.
